File: src/CormenIntoductionToAlgorithms.Tests/Sorts.fs
module Sorts

open FsCheck.Xunit
open CormenIntoductionToAlgorithms.Sorts

let private isSorted (array:int[]) =
    let rec loop current =
        let next = current + 1
        if next >= array.Length
        then true
        elif array.[current] <= array.[next]
        then loop next
        else false
    loop 0

[<Property>]
let ``Insertion sort`` array =
    InsertionSort.Sort array
    isSorted array
    
[<Property>]
let ``Selection sort`` array =
    SelectionSort.Sort array
    isSorted array

[<Property>]
let ``Bubble sort`` array =
    BubbleSort.Sort array
    isSorted array

[<Property>]
let ``Merge sort`` array =
    MergeSort.Sort array
    isSorted array

[<Property>]
let ``Heap sort`` array =
    HeapSort.Sort array
    isSorted array

File: src/CormenIntoductionToAlgorithms.Tests/Sorts.cs
using CormenIntoductionToAlgorithms.Sorts;
using FsCheck.Xunit;

namespace CormenIntoductionToAlgorithms.Tests;

public class Sorts
{
    [Property]
    public void InsertionSortTest(int[] array)
    {
        CheckSorting(array, InsertionSort.Sort);
    }

    [Property]
    public void SelectionSortTest(int[] array)
    {
        CheckSorting(array, SelectionSort.Sort);
    }

    [Property]
    public void BubbleSortTest(int[] array)
    {
        CheckSorting(array, BubbleSort.Sort);
    }

    [Property]
    public void MergeSortTest(int[] array)
    {
        CheckSorting(array, MergeSort.Sort);
    }

    [Property]
    public void HeapSortTest(int[] array)
    {
        CheckSorting(array, HeapSort.Sort);
    }

    private static void CheckSorting(int[] array, Action<int[]> sorter)
    {
        sorter(array);
        AssertSorted(array);
    }

    private static void AssertSorted(IReadOnlyList<int> list)
    {
        if (list.Count <= 1)
        {
            Assert.True(true);
            return;
        }

        var current = list[0];
        for (int i = 1; i < list.Count; i++)
        {
            var element = list[i];
            if (element >= current)
            {
                current = element;
            }
            else
            {
                Assert.Fail($"Отличается в индексе [{i-1}:{i}]");
            }
        }

        Assert.True(true);
    }
}
